File: typescript/dhondt.ts
interface Party {
    name: string;
    votes: number;
    seats: number;
}

interface Quotient {
    partyIndex: number;
    value: number;
}


function getRandomInt(min: number, max: number): number {
    return Math.floor(Math.random() * (max - min + 1)) + min;
}

function getRandomName(length: number): string {
    const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz';
    let name = '';
    for (let i = 0; i < length; i++) {
        name += characters.charAt(Math.floor(Math.random() * characters.length));
    }
    return name;
}

function generateRandomParties(numParties: number, maxVotes: number): Party[] {
    const parties: Party[] = [];
    for (let i = 0; i < numParties; i++) {
        const nameLength = getRandomInt(1, 10);
        const partyName = getRandomName(nameLength);
        const votes = getRandomInt(0, maxVotes);
        parties.push({ name: partyName, votes: votes, seats: 0 });
    }
    return parties;
}

function dHondt(parties: Party[], totalSeats: number): Party[] {
    // Initialize seats for each party to 0
    parties.forEach(party => {
        party.seats = 0;
    });

    let quotients: Quotient[] = [];

    // Generate all quotients
    for (let i = 0; i < parties.length; i++) {
        for (let j = 1; j <= totalSeats; j++) {
            quotients.push({
                partyIndex: i,
                value: parties[i].votes / j
            });
        }
    }

    // Sort quotients in descending order
    quotients.sort((a, b) => b.value - a.value);

    // Allocate seats based on the highest quotients
    for (let i = 0; i < totalSeats; i++) {
        let partyIndex = quotients[i].partyIndex;
        parties[partyIndex].seats++;
    }

    return parties;
}

const numParties = 1000;
const maxVotes = 1000;
const parties: Party[] = generateRandomParties(numParties, maxVotes);

const totalSeats = 500;
const result = dHondt(parties, totalSeats);
console.log(result);

File: lua/dhondt.lua
-- Define Party and Quotient tables (similar to TypeScript interfaces)
Party = {}
Party.__index = Party

function Party:new(name, votes, seats)
    local party = {
        name = name,
        votes = votes,
        seats = seats
    }
    setmetatable(party, Party)
    return party
end

Quotient = {}
Quotient.__index = Quotient

function Quotient:new(partyIndex, value)
    local quotient = {
        partyIndex = partyIndex,
        value = value
    }
    setmetatable(quotient, Quotient)
    return quotient
end

-- Function to get a random integer
function getRandomInt(min, max)
    return math.floor(math.random() * (max - min + 1)) + min
end

-- Function to get a random name of given length
function getRandomName(length)
    local characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz'
    local name = ''
    for i = 1, length do
        local index = math.floor(math.random() * #characters) + 1
        name = name .. characters:sub(index, index)
    end
    return name
end

-- Function to generate random parties
function generateRandomParties(numParties, maxVotes)
    local parties = {}
    for i = 1, numParties do
        local nameLength = getRandomInt(1, 10)
        local partyName = getRandomName(nameLength)
        local votes = getRandomInt(0, maxVotes)
        table.insert(parties, Party:new(partyName, votes, 0))
    end
    return parties
end

-- d'Hondt method function
function dHondt(parties, totalSeats)
    -- Initialize seats for each party to 0
    for _, party in ipairs(parties) do
        party.seats = 0
    end

    local quotients = {}

    -- Generate all quotients
    for i = 1, #parties do
        for j = 1, totalSeats do
            table.insert(quotients, Quotient:new(i, parties[i].votes / j))
        end
    end

    -- Sort quotients in descending order
    table.sort(quotients, function(a, b) return a.value > b.value end)

    -- Allocate seats based on the highest quotients
    for i = 1, totalSeats do
        local partyIndex = quotients[i].partyIndex
        parties[partyIndex].seats = parties[partyIndex].seats + 1
    end

    return parties
end

-- Main script
local numParties = 1000
local maxVotes = 1000
local parties = generateRandomParties(numParties, maxVotes)

local totalSeats = 500
local result = dHondt(parties, totalSeats)

for _, party in ipairs(result) do
    print(party.name .. ": " .. party.votes .. " votes, " .. party.seats .. " seats")
end
